circuit_knitting/cutting/cut_finding/cut_finding.py

```diff
@@ -0,0 +1,102 @@
+# This code is a Qiskit project.
+
+# (C) Copyright IBM 2024.
+
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Automatically find cut locations in a quantum circuit."""
+
+from __future__ import annotations
+
+from qiskit import QuantumCircuit
+from qiskit.circuit import CircuitInstruction
+from typing import cast
+
+from .optimization_settings import OptimizationSettings
+from .quantum_device_constraints import DeviceConstraints
+from .disjoint_subcircuits_state import DisjointSubcircuitsState
+from .circuit_interface import SimpleGateList
+from .lo_cuts_optimizer import LOCutsOptimizer
+from .cco_utils import qc_to_cco_circuit
+from ..instructions import CutWire
+from ..cutting_decomposition import cut_gates
+
+
+def find_cuts(
+    circuit: QuantumCircuit,
+    optimization: OptimizationSettings | dict[str, str | int],
+    constraints: DeviceConstraints | dict[str, int],
+) -> QuantumCircuit:
+    """
+    Find cut locations in a circuit, given optimization settings and QPU constraints.
+
+    Args:
+        circuit: The circuit to cut
+        optimization: Settings for controlling optimizer behavior. Currently,
+            only a best-first optimizer is supported. For a list of supported
+            optimization settings, see :class:`.OptimizationSettings`.
+        constraints: QPU constraints used to generate the cut location search space.
+            For information on how to specify QPU constraints, see :class:`.DeviceConstraints`.
+
+    Returns:
+        A circuit containing :class:`.BaseQPDGate` instances. The subcircuits
+        resulting from cutting these gates will be runnable on the devices
+        specified in ``constraints``.
+    """
+    circuit_cco = qc_to_cco_circuit(circuit)
+    interface = SimpleGateList(circuit_cco)
+
+    if isinstance(optimization, dict):
+        opt_settings = OptimizationSettings.from_dict(optimization)
+    else:
+        opt_settings = optimization
+
+    # Hard-code the optimization type to best-first
+    opt_settings.set_engine_selection("CutOptimization", "BestFirst")
+
+    if isinstance(constraints, dict):
+        constraint_settings = DeviceConstraints.from_dict(constraints)
+    else:
+        constraint_settings = constraints
+
+    # Hard-code the optimizer to an LO-only optimizer
+    optimizer = LOCutsOptimizer(interface, opt_settings, constraint_settings)
+
+    # Find cut locations
+    opt_out = optimizer.optimize()
+
+    wire_cut_actions = []
+    gate_ids = []
+
+    opt_out = cast(DisjointSubcircuitsState, opt_out)
+    opt_out.actions = cast(list, opt_out.actions)
```

Comment:
What error were you seeing that caused you to have to cast `opt_out` here?

Reply:
I think there was a mypy error here, because of the fact that `opt_out` is also allowed to be `None`.